core/store/src/contract.rs

```diff
@@ -24,61 +24,145 @@ pub struct ContractStorage {
             }
         }
 
-        let contract_code = match self.storage.retrieve_raw_bytes(&code_hash) {
+        match self.storage.retrieve_raw_bytes(&code_hash) {
             Ok(raw_code) => Some(ContractCode::new(raw_code.to_vec(), Some(code_hash))),
             Err(_) => None,
-        };
-
-        if contract_code.is_some() {
-            let mut guard = self.storage_reads.lock().expect("no panics");
-            guard.as_mut().expect("must not be called after finalize").insert(CodeHash(code_hash));
         }
+    }
+
+    /// Records an uncommitted call to a contract by code-hash.
+    ///
+    /// This is used to capture the contracts that are called when applying a chunk.
+    /// Calling `rollback` clears this recording.
+    pub fn record_call(&self, code_hash: CryptoHash) {
+        let mut tracker = self.tracker.lock().expect("no panics");
+        tracker.call(code_hash.into());
+    }
+
+    /// Stores the contract code as an uncommitted deploy.
+    ///
+    /// Subsequent calls to `get` will return the code that was stored here.
+    /// Calling `rollback` clears this uncommitted deploy.
+    pub fn record_deploy(&self, code: ContractCode) {
+        let mut tracker = self.tracker.lock().expect("no panics");
+        tracker.deploy(code);
+    }
 
-        contract_code
+    /// Commits the uncommitted recording of contract calls and deployments.
+    ///
+    /// This adds the uncommitted calls and deployments to the committed list and clears the uncommitted list.
+    /// Note that there can be multiple calls to `commit`.
+    pub(crate) fn commit(&mut self) {
+        let mut tracker = self.tracker.lock().expect("no panics");
+        tracker.commit();
     }
 
-    pub fn store(&self, code: ContractCode) {
-        let mut guard = self.uncommitted_deploys.lock().expect("no panics");
-        guard.insert(*code.hash(), code);
+    /// Rolls back the uncommitted recording of contract calls and deployments.
+    pub(crate) fn clear_uncommitted(&mut self) {
+        let mut tracker = self.tracker.lock().expect("no panics");
+        tracker.clear_uncommitted();
     }
 
-    /// Destructs the ContractStorage and returns the list of storage reads.
+    /// Finalizes this instance and returns the list of committed contract calls and deployments collected so far.
+    ///
+    /// This should be called when there will be no further deployments or calls to contracts.
     pub(crate) fn finalize(self) -> ContractStorageResult {
-        let mut guard = self.storage_reads.lock().expect("no panics");
-        // TODO(#11099): Change `replace` to `take` after investigating why `get` is called after the TrieUpdate
-        // is finalizing in the yield-resume tests.
-        ContractStorageResult {
-            contract_accesses: guard.replace(HashSet::new()).unwrap().into_iter().collect(),
-        }
+        let tracker = self.tracker.lock().expect("no panics");
```

Comment:
`self.tracker.into_inner` gives you the contained value by value, so you don't need to clone the inner structures.